lib/utils/log/slog_handler.go

```diff
@@ -0,0 +1,386 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"runtime"
+	"slices"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// TraceLevel is the logging level when set to Trace verbosity.
+const TraceLevel = slog.LevelDebug - 1
+
+// SlogTextHandler is a [slog.Handler] that outputs messages in a textual
+// manner as configured by the Teleport configuration.
+type SlogTextHandler struct {
+	level          slog.Leveler
+	enableColors   bool
+	component      string
+	preformatted   []byte   // data from WithGroup and WithAttrs
+	unopenedGroups []string // groups from WithGroup that haven't been opened
+	mu             *sync.Mutex
```

Comment:
We should add a comment to `mu` to state this is protecting `out` and not the rest of the fields and that it's intentional that mu is copied to child handlers. IIRC there was actually a bug in the original exp/slog implementation where a new mutex was created during the creation of child handlers and this lead to garbled logs - whilst we've not repeated that bug here - I think it's trippy enough that a sentence or two is justified.